lib/worker/OperationQueue.js

```diff
@@ -0,0 +1,363 @@
+// the number concurrent I/O operations. The number 25 was arbitrarily picked.
+const limit = pLimit(25);
+// Controls the maximum number of documents to pull from the database and cache
+// locally. The Math.max(DEFAULT_MAX_CACHED_QUEUE_SIZE, MAX_OPS_PER_REG_EVENT)
+// will be the final value used to limit the number of documents pulled from
+// the database.
+const DEFAULT_MAX_CACHED_QUEUE_SIZE = 250;
+
+// Error Messages
+const OPERATIONS_REMOVAL_ERROR = 'Failed to remove operations from queue.';
+const OPERATION_VALIDATION_ERROR = 'Non-critical operation validation error. ' +
+  'Invalid head of operation queue.';
+
+module.exports = class OperationQueue {
+  constructor({worker}) {
+    // continuity worker
+    this.worker = worker;
+    const {ledgerNode} = worker;
+    // ledger node instance
+    this.ledgerNode = ledgerNode;
+    // ledger node's id
+    this.ledgerNodeId = ledgerNode.id;
+    // operation queue collection
+    this.collection = database.collections.continuity2017_operation_queue;
+    // cached chunk of operations to return for a regular event
+    this.chunk = null;
+    // tracks whether or not there are more operations after the next chunk
+    this.hasMore = false;
+    // cached queue from the database
+    this.queue = null;
+    // index of the head of the cached queue of operations
+    this.head = 0;
+    // count of possible operations to be added in the chunk
+    this.opCount = 0;
+    // list of operations to be removed
+    this.removedOperations = [];
+    // caches operation removal promise
+    this.removing = null;
+    // caches database empty check promise
+    this.databaseEmpty = null;
+    // caches fetch record promise
+    this.fetchRecords = null;
+    // constants
+    this.MAX_OPERATIONS = _continuityConstants.events.maxOperations;
+    this.MAX_CACHED_QUEUE_SIZE = Math.max(
+      DEFAULT_MAX_CACHED_QUEUE_SIZE,
+      this.MAX_OPERATIONS);
+  }
+
+  /**
+   * Returns whether or not there is another chunk of operations that can
+   * be put into a local regular event. This MUST NOT be called concurrently.
+   *
+   * @return a Promise that resolves to `true` or `false`.
+   */
+  async hasNextChunk() {
+    // if opCount is non-zero then the function has previously been called and
+    // the chunk was not popped from the queue
+    if(this.opCount > 0) {
+      return true;
+    }
+
+    const {ledgerNodeId} = this;
+    // wait for removal of previous chunk before continuing
+    if(this.removing) {
+      await this.removing.catch(e => {
+        logger.error(OPERATIONS_REMOVAL_ERROR, {ledgerNodeId, error: e});
+      });
+    }
+
+    // if the queue is not initialized or the head of the queue is at the end
+    // of the queue, then no next chunk of operations cached yet, so create one
+    // that will be cached when `getNextChunk` is called and its operations are
+    // retrieved
+    if(!(Array.isArray(this.queue) && (this.head + 1) < this.queue.length)) {
```

Comment:
I think this is clearer for checking whether "head is before the end of the queue".

```suggestion
    const end = this.queue.length - 1;
    if(!(Array.isArray(this.queue) && this.head < end)) {
```

Reply:
Also, this feels a little OBO prone since we're not just advancing the head to the length of the queue, but I'm going to ignore that. To be clear, it seems odd to me to say that when the head is pointing to the last element in the queue we will reset it to `0` ... what about that last element?